Document mx check and mx show-version (#4721)
mendix/docs: content/en/docs/refguide/general/mx-command-line-tool.md

## 1 Introduction

The **mx tool** is a Windows and Linux command-line tool that can be used to perform various actions on a Mendix app.

## 2 Location

Mendix Studio Pro comes with the mx command-line tool. The executable `mx.exe` file can be found in the same folder that contains `studiopro.exe` (for example, *C:\Program Files\Mendix\9.12.2.44241\modeler\mx.exe*).

## 3 mx Tool Options

The mx tool performs the commands described below.

### 3.1 mx check Command

> The `mx check` command checks the app MPR file for issues such as Errors, Warnings, Deprecations or Performance recommendations.

Reviewer: I assume this only works if the MPR file is the same version as mx.exe?
If so, I think we should say so.
And does this work for all versions of Mx9, or only from a particular release?

Author: I will update.

notes:
- mx check == 9.10
- mx show-version == 9.4
- mx convert = (has always been in SP9)

future:
- mx check --performance ==  9.16

Author: done

| `--warnings` | `-w` | Include warnings in the output |
| `--deprecations` | `-d` | Include deprecations in the output |

{{% alert color="info" %}}
Errors in the MPR are always reported.
{{% /alert %}}

For `INPUT`, enter a single *.mpr* file.

#### 3.1.2 Examples

Examples of commands are described in the table below:

| Example | Result |
| --- | --- |
| `mx check --help` | Displays the help text for the check command. |
| `mx check C:\MxProjects\App-main\App-main.mpr` | Checks the app at `C:\MxProjects\App-main\App-main.mpr` for errors. |
| `mx check C:\MxProjects\App-main\App-main.mpr --warnings --deprecations` | Checks the app at `C:\MxProjects\App-main\App-main.mpr` for errors, warnings and deprecations. |

#### 3.1.3 Return Codes

Return codes are described in the table below:

| Return Code | Description |
| --- | --- |
| 0 | No issues found. |
| 1 | Errors were found. |
| 2 | Warnings were found. |
| 4 | Deprecations were found. |

Those values are logically OR combined to indicate when there are a mix of errors, warnings, deprecations.

For example:
- 3 if errors and warnings found
- 7 if errors, warnings, and deprecations found

> ### 3.2 mx convert Command [version 9+]

Reviewer: I think we can remove the version information as it is in all Mx 9 versions?
```suggestion
### 3.2 mx convert Command
```

> The `mx convert` command converts the app(s) MPK file(s) to a specific Studio Pro version. For example, if you are using the mx command-line tool for Mendix version 9.12.2.44241, then `mx convert` will convert the app to that version. 

Reviewer: Why does this work on the MPK file when all the others work on the MPR file?
I think this could cause confusion and errors.
Studio Pro also works on the MPR file when it does a conversion.

Author: hm my understanding is that `show-version` and `check` are run on apps that are already in use on the user's machine (and any conversion has already been done). So they are in the 'extracted' state which has an MPR file.

Whereas `convert` is for some app that is not yet in use, and needs to be converted.
`convert` takes a MPK file, or else a folder.

If we want to change how this behaves, that would be some future work 🤔  not related to this PR.

Reviewer: Thanks for the explanation. I've changed the text to highlight more that it is MPKs and not MPRs. If we get lots of support calls I can add a big banner, but I'll leave it for now.

The mx tool enables the options described below.

### 3.1 mx check Command

The `mx check` command checks the app for issues such as Errors, Warnings, Deprecations or Performance recommendations.

The input is an MPR file.

#### 3.1.1 Usage

Use the following command pattern for `mx check`:

> `mx convert [OPTIONS] INPUT`

Reviewer: mx check, surely.
```suggestion
`mx check [OPTIONS] INPUT`
```

Author: done

* When using the `--in-place` option, the `INPUT...` folder will also be used as the `OUTPUT` folder, so you do not need to specify a separate `OUTPUT` folder

#### 3.2.2 Examples

Examples of commands are described in the table below:

| Example | Result |
| --- | --- |
| `mx convert --in-place C:\MxProjects\App-main` | Converts the app in folder `C:\MxProjects\App-main` to the specific Studio Pro version which the mx tool is bundled with. |
| `mx convert C:\Mendix\App1.mpk C:\Mendix\App2.mpk C:\Mendix\ConvertedProjects\` | Converts the *App1.mpk* and *App2.mpk* app packages that are in the *C:\Mendix\* folder and puts the results in the `C:\Mendix\ConvertedProjects\` folder. |
| `mx convert --skip-error-check C:\Mendix\Packages\ C:\Mendix\ConvertedPackages\` | Converts all app packages in the `C:\Mendix\Packages\` folder to the `C:\Mendix\ConvertedPackages\` folder without checking for errors. |

#### 3.2.3 Return Codes

Return codes are described in the table below:

| Return Code | Description |
| --- | --- |
| 0 | The conversion was successful. |
| 1 | An internal error occurred. |
| 2 | There is something wrong with the command-line options. |
| 3 | Converting failed. |

> ### 3.3 mx create-project Command [version 9+]

Reviewer: ```suggestion
### 3.3 mx create-project Command
```

| 2 | There is something wrong with the command-line options. |
| 3 | Converting failed. |

> ### 3.3 mx create-project Command

Reviewer: Should this be deprecated and replaced with
`mx create-app`
?

Author: Possibly - that would be for SOLP team

Author: (unless I can do in some future crafting ... )

| --- | --- | --- |
| `app-name` | App | Assigns the specified app name to the app. |
| `output-dir` | Current directory | The directory in which to create the app. |

> | `language-code` | Optional | The default language of the app. | 

Reviewer: Is this really completely optional, or does it default to en-US?
What happens in Studio Pro if there is no default language set?

Author: In the source code there is a hard-coded default `en_US`

Author: So, yes it is really optional.   

default = `en_US`

Reviewer: I think putting en_US is more accurate. The first example shows that you don't need to put any options.
```suggestion
| `language-code` | en_US | The default language of the app. | 
```

| `sprintr-app-id` | Optional | Associates the app [feedback features](/developerportal/collaborate/feedback/) with the provided [Developer Portal app](/developerportal/#my-apps). |

`TEMPLATE-MPK-FILE` is an optional path to a Mendix app package (*.mpk*) file. If this argument is omitted, the app is created with a default empty project template.

> #### 3.3.2 Examples

Reviewer: Would be nice to have an example with a `sprintr-app-id` so customers can see the format of this.

We also don't use Sprintr in customer facing documentation, so can you think of a better name for this option? I assume that it isn't going to be the same as a Team Server app id?

Author: I believe `sprintr-app-id` is a Team Server app ID.

I will check  ...

Author: yes they are the same.

The value is a GUID. 
When accessing the app portal (on sprintr OR on Team Server) it can be seen in the browser's URL

sprintr example:
https://sprintr.home.mendix.com/link/buzz/1a428ea7-b00e-4166-9b23-20b7bf88a70d

team server example:

https://teamserver.sprintr.com/1a428ea7-b00e-4166-9b23-20b7bf88a70d

Author: done

> | `mx create-project --app-name "MyFirstApp" --output-dir "C:/Projects/MyFirstApp"` | Creates an app named `MyFirstApp` in the *C:/Projects/MyFirstApp* folder using all the default parameters. |

Reviewer: The options here are quoted. In `mx convert` they are not quoted. Should we be consistent and always quote the parameters?

Author: hm normally the quoting is not needed.

so maybe it makes sense to remove the quotes ?

Author: related: most of the commands use `Windows` folder separators like `\`
but `mx create-project` uses Unix style.

suggest: `mx create-project` text be changed to be consistent with the other commands

what do you think?

Reviewer: Great catch.
Yes, I think we should be consistent and use either all Windows style or all Unix style.
And perhaps remove the quotes - but I assume it will work if you quote directory paths with spaces. I think we can assume that customers know this.

Author: done
